android/src/app/organicmaps/car/util/UiHelpers.java

```diff
@@ -50,13 +60,89 @@ public static MapController createMapController(@NonNull CarContext context, @No
   }
 
   @NonNull
-  public static Action createSettingsAction(@NonNull CarContext context, @NonNull SurfaceRenderer surfaceRenderer)
+  public static Action createSettingsAction(@NonNull BaseMapScreen mapScreen, @NonNull SurfaceRenderer surfaceRenderer)
   {
+    final CarContext context = mapScreen.getCarContext();
     final CarIcon iconSettings = new CarIcon.Builder(IconCompat.createWithResource(context, R.drawable.ic_settings)).build();
 
-    return new Action.Builder().setIcon(iconSettings).setOnClickListener(
-        () -> context.getCarService(ScreenManager.class).push(new SettingsScreen(context, surfaceRenderer))
-    ).build();
+    return new Action.Builder().setIcon(iconSettings).setOnClickListener(() -> {
+      // Action.onClickListener for the Screen A maybe called even if the Screen B is shown now.
+      // We need to check it
+      // This may happen when we use PopToRootHack:
+      //   * ScreenManager.popToRoot()
+      //   * The root screen (A) is shown for a while
+      //   * User clicks on some action
+      //   * ScreenManager.push(new Screen())
+      //   * New screen (B) is displayed now
+      //   * Action.onClickListener is called for action from root screen (A)
```

Comment:
Have you checked the demo with the example of this thing on `0:08 - 0:10`?
I'm opening a new screen and still see the old one for some time. I'm trying to click buttons on the old screen and nothing happens. That's how it works.

I can't say whether this is a bug or a "feature" but here are my thoughts on how it works.
I guess they have to separate queues for pushing/popping a Screen and for handling button clicks. And they are not synchronized.
How it works:
We push a new screen to the stack. But it's not displayed instantly. Some processing goes maybe on a separate thread and we're still able to interact with the interface. So, let's say it just goes to a screen queue.
We click the button on the old screen (because we can interact with the interface before a new screen is displayed). And again, it's not processed instantly. It goes to the buttons queue.
Now, the main thread is free and a new screen is displayed. The screen queue is empty and the button queue is being processed. 
There is an action stored in the button queue (click on the old screen). And it's applied even if the new screen is displayed. 

In simple use-case that should never happen but we have the `PopToRoot` hack that brings some troubles 😄

Reply:
OK, I see. Thanks for explanation. Let's keep this hack in favor of better user experience.